src/main/flight/imu.c

```diff
@@ -529,3 +526,73 @@ void imuSetHasNewData(uint32_t dt)
 }
 #endif
 
+void imuQuaternionComputeProducts(quaternion *quat) {
+    quat->ww = quat->w * quat->w;
+    quat->wx = quat->w * quat->x;
+    quat->wy = quat->w * quat->y;
+    quat->wz = quat->w * quat->z;
+    quat->xx = quat->x * quat->x;
+    quat->xy = quat->x * quat->y;
+    quat->xz = quat->x * quat->z;
+    quat->yy = quat->y * quat->y;
+    quat->yz = quat->y * quat->z;
+    quat->zz = quat->z * quat->z;
+}
+
+bool imuQuaternionHeadfreeDislocationSet(void) {
+    if((fabsf(attitude.values.roll/10.0f) < 45.0f)  && (fabsf(attitude.values.pitch/10.0f) < 45.0f)) {
+        const float roll = 0;
+        const float pitch = 0;
+        const float yaw = -atan2f((+2.0f * (q.wz + q.xy)), (+1.0f - 2.0f * (q.yy + q.zz)));
+
+        const float cosRoll2 = cos(roll/2);
+        const float cosPitch2 = cos(pitch/2);
+        const float cosYaw2 = cos(yaw/2);
+        const float sinRoll2 = sin(roll/2);
+        const float sinPitch2 = sin(pitch/2);
+        const float sinYaw2 = sin(yaw/2);
+        const float cosPitch2cosYaw2 = cosPitch2 * cosYaw2;
+        const float sinPitch2sinYaw2 = sinPitch2 * sinYaw2;
+
+        dislocation.w = cosRoll2 * cosPitch2cosYaw2 + sinRoll2 * sinPitch2sinYaw2 ;
+        dislocation.x = sinRoll2 * cosPitch2cosYaw2 - cosRoll2 * sinPitch2sinYaw2 ;
+        dislocation.y = cosRoll2 * sinPitch2 * cosYaw2 + sinRoll2 * cosPitch2 * sinYaw2 ;
+        dislocation.z = cosRoll2 * cosPitch2 * sinYaw2 - sinRoll2 * sinPitch2 * cosYaw2;
+
+        imuQuaternionComputeProducts(&dislocation);
+
+        return(true);
+    } else {
+        return(false);
+    }
+}
+
+void imuQuaternionMultiplication(quaternion *q1, quaternion *q2, quaternion *result) {
+    const float A = (q1->w + q1->x)*(q2->w + q2->x);
+    const float B = (q1->z - q1->y)*(q2->y - q2->z);
+    const float C = (q1->w - q1->x)*(q2->y + q2->z);
+    const float D = (q1->y + q1->z)*(q2->w - q2->x);
+    const float E = (q1->x + q1->z)*(q2->x + q2->y);
+    const float F = (q1->x - q1->z)*(q2->x - q2->y);
+    const float G = (q1->w + q1->y)*(q2->w - q2->z);
+    const float H = (q1->w - q1->y)*(q2->w + q2->z);
+
+    result->w = B + (-E - F + G + H)/2;
+    result->x = A - (E + F + G + H)/2;
+    result->y = C + (E - F + G - H)/2;
+    result->z = D + (E - F - G + H)/2;
+
+    imuQuaternionComputeProducts(result);
+}
+
+void imuQuaternionHeadfreeTransformVectorEarthToBody(t_fp_vector_def *v) {
+    imuQuaternionMultiplication(&dislocation, &q, &headfree);
+
+    const float x = (headfree.ww + headfree.xx - headfree.yy - headfree.zz) * v->X + 2*(headfree.xy + headfree.wz) * v->Y + 2*(headfree.xz - headfree.wy) * v->Z;
```

Comment:
You can convert `headfree` to rotation matrix and use it here.
Or rotate attitude matrix around Z (https://en.wikipedia.org/wiki/Rotation_matrix#Basic_rotations), it may be faster than using quternion